File: agent.py
"""Agent IA local d'interpretation de l'epellation ASL (100% gratuit, hors-ligne).

Boucle d'agent : PERCEPTION (flux de lettres) -> MEMOIRE (tampon, vocabulaire FR+EN,
abreviations) -> RAISONNEMENT (nettoyage, segmentation, correction, expansion des
abreviations, detection de langue) -> ACTION (texte interprete, suggestions, journal),
avec declenchement autonome sur pause (desactivable).
"""
import os, time, threading, difflib, re
from functools import lru_cache

def _read_words(fn):
    out = []
    if os.path.exists(fn):
        with open(fn) as f:
            out = f.read().split()
    return out

def _load_vocab():
    fr = _read_words("assets/vocab_fr.txt")
    en = _read_words("assets/vocab_en.txt")
    words, rank = [], {}
    for lst in (fr, en):
        for i, w in enumerate(lst):
            if len(w) == 1 and w not in ("A", "I"):
                continue
            if w not in rank:
                rank[w] = i
                words.append(w)
    by_len = {}
    for w in words:
        by_len.setdefault(len(w), []).append(w)
    return words, rank, by_len, set(fr), set(en)

VOCAB, RANK, BY_LEN, FR_SET, EN_SET = _load_vocab()
VOCAB_SET = set(VOCAB)

# --- Abreviations conventionnelles de l'epellation (anglais + francais) ---
ABBREV_EN = {
    "U": "you", "R": "are", "UR": "your", "YR": "your", "Y": "why",
    "N": "and", "B": "be", "C": "see", "K": "ok", "OK": "ok",
    "THX": "thanks", "TY": "thank you", "PLS": "please", "PLZ": "please",
    "BC": "because", "BCS": "because", "W": "with", "WO": "without",
    "TMR": "tomorrow", "TMRW": "tomorrow", "TDY": "today", "RN": "right now",
    "MSG": "message", "NVM": "nevermind", "IDK": "i dont know",
    "BRB": "be right back", "OMW": "on my way", "ILY": "i love you",
}
ABBREV_FR = {
    "BJR": "bonjour", "SLT": "salut", "CC": "coucou", "CV": "ca va",
    "MRC": "merci", "BCP": "beaucoup", "DSL": "desole", "STP": "s'il te plait",
    "SVP": "s'il vous plait", "PK": "pourquoi", "PCQ": "parce que",
    "AJD": "aujourd'hui", "MTN": "maintenant", "TT": "tout", "TLM": "tout le monde",
    "JSP": "je sais pas", "JTM": "je t'aime", "QQN": "quelqu'un",
    "QQCH": "quelque chose", "RDV": "rendez-vous", "STV": "si tu veux",
}
ABBREV = {}
ABBREV.update(ABBREV_EN)
ABBREV.update(ABBREV_FR)

@lru_cache(maxsize=4096)
def _close(word):
    cands = []
    for L in (len(word) - 1, len(word), len(word) + 1):
        cands.extend(BY_LEN.get(L, []))
    m = difflib.get_close_matches(word, cands, n=1, cutoff=0.78)
    return m[0] if m else None

@lru_cache(maxsize=8192)
def _word_cost(w):
    BREAK = 1.6
    if w in RANK:
        c = 0.5 + RANK[w] / 8000.0
        if len(w) == 1:
            c += 2.2
        elif len(w) == 2 and RANK[w] > 1500:
            c += 2.5
        return c + BREAK
    if 4 <= len(w) <= 9:
        m = _close(w)
        if m:
            return 2.0 + RANK[m] / 8000.0 + 3.2 + BREAK
    return 4.0 + 1.6 * len(w) + BREAK

def collapse_repeats(s):
    return re.sub(r"(.)\1{2,}", r"\1\1", s)

def segment(s):
    n = len(s)
    if n == 0:
        return ""
    cost = [0.0] + [float("inf")] * n
    back = [0] * (n + 1)
    for i in range(1, n + 1):
        for j in range(max(0, i - 16), i):
            c = cost[j] + _word_cost(s[j:i])
            if c < cost[i]:
                cost[i] = c; back[i] = j
    out, i = [], n
    while i > 0:
        out.append(s[back[i]:i]); i = back[i]
    return " ".join(reversed(out))

def correct_word(w):
    if w in VOCAB_SET or len(w) < 3:
        return w, False
    m = _close(w)
    if m:
        return m, True
    return w, False

def suggest(prefix, k=3):
    p = re.sub(r"[^A-Z]", "", prefix.upper())
    if len(p) < 2:
        return []
    out = []
    for w in VOCAB:
        if w.startswith(p) and w != p:
            out.append(w)
            if len(out) >= k:
                break
    return out

def _detect_lang(words):
    fr = sum(1 for w in words if w.upper() in FR_SET and w.upper() not in EN_SET)
    en = sum(1 for w in words if w.upper() in EN_SET and w.upper() not in FR_SET)
    if fr > en:
        return "francais"
    if en > fr:
        return "anglais"
    return None

def interpret(raw):
    """Pipeline de raisonnement. Retourne (texte_interprete, journal des decisions)."""
    journal = []
    s = re.sub(r"[^A-Za-z ]", "", raw).upper().strip()
    if not s:
        return "", ["(rien a interpreter)"]
    t = collapse_repeats(s)
    if t != s:
        journal.append("Repetitions supprimees : " + s + " -> " + t)

    def expand_or_correct(w):
        if w in ABBREV:
            rep = ABBREV[w]
            journal.append("Abreviation : " + w + " -> " + rep)
            return rep.split()
        cw, changed = correct_word(w)
        if changed:
            journal.append("Correction : " + w + " -> " + cw)
        return [cw]

    words = []
    for chunk in t.split():
        if chunk in ABBREV:
            rep = ABBREV[chunk]
            journal.append("Abreviation : " + chunk + " -> " + rep)
            words.extend(rep.split())
            continue
        if chunk in VOCAB_SET:
            words.append(chunk); continue
        seg = segment(chunk)
        if seg != chunk:
            journal.append("Segmentation : " + chunk + " -> " + seg)
        for w in seg.split():
            words.extend(expand_or_correct(w))

    lang = _detect_lang(words)
    if lang:
        journal.append("Langue detectee : " + lang)

    out = [w.lower() for w in words]
    if lang == "anglais":
        out = ["I" if w == "i" else w for w in out]     # 'I' majuscule en anglais
    final = " ".join(out).strip()
    final = (final[:1].upper() + final[1:]) if final else final
    if not journal:
        journal.append("Texte deja coherent, aucune correction necessaire.")
    return final, journal

class InterpretingAgent:
    """Observe le flux, detecte les pauses, decide (si le mode auto est actif), agit."""
    def __init__(self, pause_s=5.0):
        self.pause_s = pause_s
        self.auto = True
        self.lock = threading.Lock()
        self.last_hand_t = time.time()
        self.last_interpreted_raw = None
        self.interpretation = ""
        self.journal = []
        self.suggestions = []

    def observe(self, raw_sentence, hand_present):
        now = time.time()
        with self.lock:
            if not raw_sentence.strip():
                self.last_interpreted_raw = None      # texte vide -> pret pour la phrase suivante
            if hand_present:
                self.last_hand_t = now
                last_word = raw_sentence.split(" ")[-1] if raw_sentence else ""
                self.suggestions = suggest(last_word)
            elif self.auto:
                if (now - self.last_hand_t) >= self.pause_s and raw_sentence.strip() \
                        and raw_sentence != self.last_interpreted_raw:
                    self.interpretation, self.journal = interpret(raw_sentence)
                    self.last_interpreted_raw = raw_sentence

    def interpret_now(self, raw_sentence):
        with self.lock:
            self.interpretation, self.journal = interpret(raw_sentence)
            self.last_interpreted_raw = raw_sentence
            return self.interpretation

    def state(self):
        with self.lock:
            return self.interpretation, list(self.journal), list(self.suggestions)

    def consume_interpretation(self):
        """Recupere l'interpretation courante et remet l'agent en attente (mode conversation).
        Note anti-doublon : last_interpreted_raw est CONSERVE pour que le meme texte brut ne
        puisse pas re-declencher une interpretation pendant que l'interface vide le tampon ;
        il est remis a zero par observe() des que le texte redevient vide."""
        with self.lock:
            out = self.interpretation
            self.interpretation = ""
            self.journal = []
            return out

    def get_suggestions(self):
        with self.lock:
            return list(self.suggestions)

File: web/src/lib/agent.ts
// Port of agent.py: turns a raw fingerspelled stream into a clean sentence, and explains each step.
import { getCloseMatches } from "./difflib";

const ABBREV: Record<string, string> = {
  // English
  U: "you", R: "are", UR: "your", YR: "your", Y: "why", N: "and", B: "be", C: "see", K: "ok", OK: "ok",
  THX: "thanks", TY: "thank you", PLS: "please", PLZ: "please", BC: "because", BCS: "because",
  W: "with", WO: "without", TMR: "tomorrow", TMRW: "tomorrow", TDY: "today", RN: "right now",
  MSG: "message", NVM: "nevermind", IDK: "i dont know", BRB: "be right back", OMW: "on my way",
  ILY: "i love you",
  // French
  BJR: "bonjour", SLT: "salut", CC: "coucou", CV: "ca va", MRC: "merci", BCP: "beaucoup",
  DSL: "desole", STP: "s'il te plait", SVP: "s'il vous plait", PK: "pourquoi", PCQ: "parce que",
  AJD: "aujourd'hui", MTN: "maintenant", TT: "tout", TLM: "tout le monde", JSP: "je sais pas",
  JTM: "je t'aime", QQN: "quelqu'un", QQCH: "quelque chose", RDV: "rendez-vous", STV: "si tu veux",
};

export interface Interpretation { text: string; journal: string[] }

export class Vocabulary {
  readonly words: string[] = [];
  readonly rank = new Map<string, number>();
  readonly byLen = new Map<number, string[]>();
  readonly fr: Set<string>;
  readonly en: Set<string>;
  readonly all: Set<string>;

  constructor(frWords: string[], enWords: string[]) {
    for (const list of [frWords, enWords]) {
      list.forEach((w, i) => {
        if (w.length === 1 && w !== "A" && w !== "I") return;
        if (!this.rank.has(w)) { this.rank.set(w, i); this.words.push(w); }
      });
    }
    for (const w of this.words) {
      const list = this.byLen.get(w.length);
      if (list) list.push(w); else this.byLen.set(w.length, [w]);
    }
    this.fr = new Set(frWords);
    this.en = new Set(enWords);
    this.all = new Set(this.words);
  }

  static fromText(fr: string, en: string) {
    const split = (s: string) => s.split(/\s+/).filter(Boolean);
    return new Vocabulary(split(fr), split(en));
  }
}

export class Agent {
  private closeCache = new Map<string, string | null>();
  private costCache = new Map<string, number>();

  constructor(private vocab: Vocabulary) {}

  private close(word: string): string | null {
    const hit = this.closeCache.get(word);
    if (hit !== undefined) return hit;
    const cands: string[] = [];
    for (const L of [word.length - 1, word.length, word.length + 1]) cands.push(...(this.vocab.byLen.get(L) ?? []));
    const m = getCloseMatches(word, cands, 1, 0.78);
    const out = m.length ? m[0] : null;
    this.closeCache.set(word, out);
    return out;
  }

  private wordCost(w: string): number {
    const hit = this.costCache.get(w);
    if (hit !== undefined) return hit;
    const BREAK = 1.6;
    let c: number;
    const r = this.vocab.rank.get(w);
    if (r !== undefined) {
      c = 0.5 + r / 8000.0;
      if (w.length === 1) c += 2.2;
      else if (w.length === 2 && r > 1500) c += 2.5;
      c += BREAK;
    } else {
      const m = w.length >= 4 && w.length <= 9 ? this.close(w) : null;
      c = m ? 2.0 + this.vocab.rank.get(m)! / 8000.0 + 3.2 + BREAK : 4.0 + 1.6 * w.length + BREAK;
    }
    this.costCache.set(w, c);
    return c;
  }

  segment(s: string): string {
    const n = s.length;
    if (!n) return "";
    const cost = [0, ...Array(n).fill(Infinity)];
    const back = Array(n + 1).fill(0);
    for (let i = 1; i <= n; i++) {
      for (let j = Math.max(0, i - 16); j < i; j++) {
        const c = cost[j] + this.wordCost(s.slice(j, i));
        if (c < cost[i]) { cost[i] = c; back[i] = j; }
      }
    }
    const out: string[] = [];
    for (let i = n; i > 0; i = back[i]) out.push(s.slice(back[i], i));
    return out.reverse().join(" ");
  }

  correctWord(w: string): [string, boolean] {
    if (this.vocab.all.has(w) || w.length < 3) return [w, false];
    const m = this.close(w);
    return m ? [m, true] : [w, false];
  }

  suggest(prefix: string, k = 3): string[] {
    const p = prefix.toUpperCase().replace(/[^A-Z]/g, "");
    if (p.length < 2) return [];
    const out: string[] = [];
    for (const w of this.vocab.words) {
      if (w.startsWith(p) && w !== p) {
        out.push(w);
        if (out.length >= k) break;
      }
    }
    return out;
  }

  private detectLang(words: string[]): "French" | "English" | null {
    let fr = 0, en = 0;
    for (const w of words) {
      const u = w.toUpperCase();
      if (this.vocab.fr.has(u) && !this.vocab.en.has(u)) fr++;
      if (this.vocab.en.has(u) && !this.vocab.fr.has(u)) en++;
    }
    return fr > en ? "French" : en > fr ? "English" : null;
  }

  interpret(raw: string): Interpretation {
    const journal: string[] = [];
    const s = raw.replace(/[^A-Za-z ]/g, "").toUpperCase().trim();
    if (!s) return { text: "", journal: ["(nothing to interpret)"] };
    const t = collapseRepeats(s);
    if (t !== s) journal.push(`Removed repeats: ${s} -> ${t}`);

    const words: string[] = [];
    const expandOrCorrect = (w: string) => {
      if (Object.hasOwn(ABBREV, w)) {
        journal.push(`Shorthand: ${w} -> ${ABBREV[w]}`);
        words.push(...ABBREV[w].split(" "));
        return;
      }
      const [cw, changed] = this.correctWord(w);
      if (changed) journal.push(`Fixed typo: ${w} -> ${cw}`);
      words.push(cw);
    };

    for (const chunk of t.split(/\s+/).filter(Boolean)) {
      if (Object.hasOwn(ABBREV, chunk)) {
        journal.push(`Shorthand: ${chunk} -> ${ABBREV[chunk]}`);
        words.push(...ABBREV[chunk].split(" "));
        continue;
      }
      if (this.vocab.all.has(chunk)) { words.push(chunk); continue; }
      const seg = this.segment(chunk);
      if (seg !== chunk) journal.push(`Split words: ${chunk} -> ${seg}`);
      for (const w of seg.split(" ")) expandOrCorrect(w);
    }

    const lang = this.detectLang(words);
    if (lang) journal.push(`Language: ${lang}`);
    let out = words.map((w) => w.toLowerCase());
    if (lang === "English") out = out.map((w) => (w === "i" ? "I" : w));
    let text = out.join(" ").trim();
    text = text ? text[0].toUpperCase() + text.slice(1) : text;
    if (!journal.length) journal.push("Already clean, nothing to change.");
    return { text, journal };
  }
}

export function collapseRepeats(s: string): string {
  return s.replace(/(.)\1{2,}/g, "$1$1");
}

/** Watches the letter stream and interprets the sentence on its own after a pause. */
export class InterpretingAgent {
  auto = true;
  pauseS = 5;
  interpretation = "";
  journal: string[] = [];
  suggestions: string[] = [];
  private lastHandT = performance.now() / 1000;
  private lastInterpretedRaw: string | null = null;

  constructor(private agent: Agent) {}

  observe(raw: string, handPresent: boolean, now = performance.now() / 1000) {
    if (!raw.trim()) this.lastInterpretedRaw = null;
    if (handPresent) {
      this.lastHandT = now;
      const last = raw ? raw.split(" ").pop() ?? "" : "";
      this.suggestions = this.agent.suggest(last);
    } else if (this.auto && now - this.lastHandT >= this.pauseS && raw.trim() && raw !== this.lastInterpretedRaw) {
      this.setResult(this.agent.interpret(raw));
      this.lastInterpretedRaw = raw;
    }
  }

  interpretNow(raw: string) {
    this.setResult(this.agent.interpret(raw));
    this.lastInterpretedRaw = raw;
    return this.interpretation;
  }

  /** Take the current interpretation and wait for the next sentence (conversation mode). */
  consume(): string {
    const out = this.interpretation;
    this.interpretation = "";
    this.journal = [];
    return out;
  }

  private setResult(r: Interpretation) {
    this.interpretation = r.text;
    this.journal = r.journal;
  }
}
